### DISCUSSION TITLE
Different behavior from v1.1.318 to v1.1.320 RE: narrowing type Union containing generic class w.r.t. ParamSpec

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#5752] [Q&A]
Essentially I have 2 generic classes, `Parent` and `Child`. `Parent` is generic w.r.t. 1 `TypeVar`, and `Child` is generic w.r.t. 2 `TypeVar` and a `ParamSpec`. I have a `TypeAlias` that is the union of these 2 types, and a function that returns an instance of either of these 2 types based on some condition. I declare a variable with the `TypeAlias` `Union` and then assign it based on the condition. I noticed when upgrading from 1.1.318 to 1.1.320 that where assigning to the declared variable used to narrow the type to `Child`, now it is `Parent | Child`, i.e. it stopped narrowing the types. I noticed this only happens when `Child` is generic w.r.t. a `ParamSpec`. Here is some code illustrating what I'm talking about:

```python
from typing import Callable, Generic, TypeVar, Union

from typing_extensions import ParamSpec, TypeAlias


S = TypeVar("S", bound=int)
TP = ParamSpec("TP")
T = TypeVar("T", bound=int)


class Parent(Generic[S]):
    def __init__(self, s: S) -> None:
        super().__init__()
        self.s = s


############################################################################


class ChildWithoutParamSpec(Parent[S], Generic[S, T]):
    def __init__(self, s: S, t: T) -> None:
        super().__init__(s)
        self.t = t


UnionWithoutParamSpec: TypeAlias = Union[Parent, ChildWithoutParamSpec]


def without_paramspec(do_subclass: bool) -> UnionWithoutParamSpec:
    obj: UnionWithoutParamSpec
    if do_subclass:
        obj = ChildWithoutParamSpec(1, 2)
        # pyright <=1.1.318: Type of "obj" is "ChildWithoutParamSpec[Unknown, int] | ChildWithoutParamSpec[Unknown, Unknown]"
        # pyright >=1.1.320: Type of "obj" is "ChildWithoutParamSpec[Unknown, int] | ChildWithoutParamSpec[Unknown, Unknown]"
        reveal_type(obj)
    else:
        obj = Parent(1)
        reveal_type(obj)
    return obj



############################################################################


class ChildWithParamSpec(Parent[S], Generic[S, TP, T]):
    def __init__(
        self,
        s: S,
        t_callable: Callable[TP, T],
        *args: TP.args,
        **kwargs: TP.kwargs,
    ) -> None:
        super().__init__(s)
        self.t = t_callable(*args, **kwargs)


UnionWithParamSpec: TypeAlias = Union[Parent, ChildWithParamSpec]


def str_to_int(s: str) -> int:
    return int(s)


def with_paramspec(do_subclass: bool) -> UnionWithParamSpec:
    obj: UnionWithParamSpec
    if do_subclass:
        obj = ChildWithParamSpec(1, str_to_int, "1")
        # pyright <=1.1.318: Type of "obj" is "ChildWithParamSpec[Unknown, (s: str), int] | ChildWithParamSpec[Unknown, (...), Unknown]"
        # pyright >=1.1.320: Type of "obj" is "ChildWithParamSpec[Unknown, (s: str), int] | ChildWithParamSpec[Unknown, (...), int] | Parent[Unknown] | ChildWithParamSpec[Unknown, (...), Unknown]"
        reveal_type(obj)
    else:
        obj = Parent(1)
        reveal_type(obj)
    return obj

```

This would seem to be a bug to me, unless I am missing something. Any insight is much appreciated.

**Comment by erictraut** [comment]
The first thing that immediately jumps out at me is that you're defining and using type aliases in a manner that is suspect. Let's look at `UnionWithoutParamSpec`. This is a union of two generic types (`Parent` and `ChildWithoutParamSpec`). The `Parent` class expects a single type argument, and the `ChildWithoutParamSpec` expects two type arguments, but you haven't specified any type arguments. That means they are all assumed to be `Any`. Is that what you intend? If so, it would be better for you to make this explicit. In any case, you've effectively defined `UnionWithoutParamSpec` to be `Parent[Any] | ChildWithoutParamSpec[Any, Any]`.

Similarly for the `UnionWithParamSpec` type alias. You have not provided any type arguments for `Parent` or `ChildWithParamSpec`. That means these type arguments will implicitly be `Any` (for the TypeVars) and `...` (for the ParamSpec). Is that what you intend?

I recommend that you enable the `reportMissingTypeArgument` diagnostic check and fill in all of the missing type arguments to make your intent clear. (You can add the comment `# pyright: reportMissingTypeArgument=true` to the top of the file to enable this rule for just this file.) After doing this, if you still have questions about pyright's behavior, please post the updated code.

**Reply by ringohoffman** [reply]
@erictraut 

my mistake. The problem seems to persist, even with the clarified intent. Please let me know if you notice any other problems with my set up:

```python
from typing import Callable, Generic, TypeVar, Union

from typing_extensions import ParamSpec, TypeAlias


S = TypeVar("S", bound=int)
TP = ParamSpec("TP")
T = TypeVar("T", bound=int)


class Parent(Generic[S]):
    def __init__(self, s: S) -> None:
        super().__init__()
        self.s = s


############################################################################


class ChildWithoutParamSpec(Parent[S], Generic[S, T]):
    def __init__(self, s: S, t: T) -> None:
        super().__init__(s)
        self.t = t


UnionWithoutParamSpec: TypeAlias = Union[Parent[S], ChildWithoutParamSpec[S, T]]


def without_paramspec(do_subclass: bool) -> UnionWithoutParamSpec[int, int]:
    obj: UnionWithoutParamSpec[int, int]
    if do_subclass:
        obj = ChildWithoutParamSpec(1, 2)
        # pyright <=1.1.318: Type of "obj" is "ChildWithoutParamSpec[int, int]"
        # pyright >=1.1.320: Type of "obj" is "ChildWithoutParamSpec[int, int]"
        reveal_type(obj)
    else:
        obj = Parent(1)
        reveal_type(obj)
    return obj



############################################################################


class ChildWithParamSpec(Parent[S], Generic[S, TP, T]):
    def __init__(
        self,
        s: S,
        t_callable: Callable[TP, T],
        *args: TP.args,
        **kwargs: TP.kwargs,
    ) -> None:
        super().__init__(s)
        self.t = t_callable(*args, **kwargs)


UnionWithParamSpec: TypeAlias = Union[Parent[S], ChildWithParamSpec[S, TP, T]]


def str_to_int(s: str) -> int:
    return int(s)


def with_paramspec(do_subclass: bool) -> UnionWithParamSpec[int, ..., int]:
    obj: UnionWithParamSpec[int, ..., int]
    if do_subclass:
        obj = ChildWithParamSpec(1, str_to_int, "1")
        # pyright <=1.1.318: Type of "obj" is "ChildWithParamSpec[int, (s: str), int] | ChildWithParamSpec[int, (...), int]"
        # pyright >=1.1.320: Type of "obj" is "ChildWithParamSpec[int, (s: str), int] | ChildWithParamSpec[int, (...), int] | Parent[int]"
        reveal_type(obj)
    else:
        obj = Parent(1)
        reveal_type(obj)
    return obj

```

**Reply by erictraut** [reply]
Thanks for the update.

Version 1.1.321 and newer matches the behavior that you were seeing prior to 1.1.318. This was a temporary regression that wasn't caught initially by our unit tests. Apologies for any inconvenience. If you upgrade to the latest version (1.1.323), you should get the expected behavior.

**Reply by ringohoffman** [reply]
I think I do still see this issue on 1.1.323:

```console
$ pip freeze | grep pyright
pyright==1.1.323
$ pyright a.py
a.py
  a.py:35:21 - information: Type of "obj" is "ChildWithoutParamSpec[int, int]"
  a.py:38:21 - information: Type of "obj" is "Parent[int]"
  a.py:71:21 - information: Type of "obj" is "ChildWithParamSpec[int, (s: str), int] | ChildWithParamSpec[int, (...), int] | Parent[int]"
  a.py:74:21 - information: Type of "obj" is "Parent[int]"
0 errors, 0 warnings, 4 informations 
```

**Reply by erictraut** [reply]
Ah yes, I missed that. I'm going to create a bug report out of this issue and investigate further. The problem is in the logic that pyright implements for narrowing types on assignment. The logic is complex because it needs to carefully handle cases where the declared target type contains `Any` or `...`. In your example, the declared type contains `...`, and this logic is attempting to preserve this in the resulting narrowed type after the assignment.

This regression wasn't caught in any of our tests because you're doing something very unusual here. The type alias `UnionWithParamSpec: TypeAlias = Union[Parent[S], ChildWithParamSpec[S, TP, T]]` is unnecessary complex because it's equivalent to just `Parent[S]`. That's because all `ChildWithParamSpec[S, TP, T]` instances are also `Parent[S]` instances. If you simplify the type alias accordingly, you won't hit this bug.
